File: src/c/partition/s/par.c
#include <stdbool.h>

#include "small.h"

int small_sum_par(int *arr, int size, unsigned long long int index)
{
    int sum = 0;
    for (int i = 0; i < size; i++)
    {
        if (index & (1 << i))
        {
            sum += arr[size - 1 - i];
        }
    }
    return sum;
}

bool *small_par(int **arr, Partitions *p)
{
    bool *result = (bool *)malloc(p->rows * sizeof(bool));
    bool paddedResult[p->rows][128];

#pragma omp parallel default(none) shared(arr, p, result, paddedResult)
    {
#pragma omp for
        for (int i = 0; i < p->rows; i++)
        {
            result[i] = false;

            int *row = arr[i];
            int size = p->cols[i];

            unsigned long long int combs = 1 << (size - 1);
            unsigned long long int all = ((unsigned long long int)1 << size) - 1;

            int problem_sum = small_sum_par(row, size, all);
            if (problem_sum % 2 != 0)
                continue;
            int half_sum = problem_sum / 2;

            for (int j = 0; j < combs; j++)
            {
                int sum = small_sum_par(row, size, j);
                if (sum == half_sum)
                {
                    result[i] = true;
                    break;
                }
            }
        }
    }

    return result;
}

File: src/go/partition/s/par.go
package small

import (
	"sync"
	"runtime"
)

func small_sum_par(arr []int, size int, index int64) int {
	var sum int = 0

	for i := 0; i < size; i++ {
		if index&(1<<i) != 0 {
			sum += arr[i]
		}
	}

	return sum
}

func Par(arr [][]int) []int32 {
	var result []int32 = make([]int32, len(arr))

	threads := runtime.NumCPU()
	thread_problems := len(arr) / threads

	var wg sync.WaitGroup
	for w := 0; w < threads; w++ {
		wg.Add(1)

		go func(w int) {
			defer wg.Done()

			for i := 0; i <= thread_problems; i++ {
				
				var problem int = i*threads + w
				if problem >= len(arr) {
					break
				}

				result[problem] = 0
				var combs int64 = 1 << (len(arr[problem]) - 1)
				var all int64 = (1 << len(arr[problem])) - 1

				var problem_sum int = small_sum_par(arr[problem], len(arr[problem]), all)
				if problem_sum%2 != 0 {
					continue
				}
				var half_sum int = problem_sum / 2

				for j := int64(0); j < combs; j++ {
					var sum int = small_sum_par(arr[problem], len(arr[problem]), j)
					if sum == half_sum {
						result[problem] = 1
						break
					}
				}
			}
		}(w)
	}

	wg.Wait()

	return result
}
